### PR TITLE
Properly zoom to feature for single point selected

### PR DESCRIPTION
**Pull request by 3nids**
fix #9160
this will fetch the canvas content and combine selected point with closest point and scale it 5 times

just before the line fix for @andreasneumann

### CODE REVIEW

**Comment by andreasneumann** [comment]
Yeah  - much appreciated!

File: src/gui/qgsmapcanvas.cpp [hunk 1]
```diff
@@ -990,8 +990,39 @@ void QgsMapCanvas::zoomToSelected( QgsVectorLayer *layer )
...
 
   rect = mapSettings().layerExtentToOutputExtent( layer, rect );
+
+  // zoom in if point cannot be distinguish from others
+  // also check that rect is empty, as it might not in case of multi points
+  if ( layer->geometryType() == QgsWkbTypes::PointGeometry && rect.isEmpty() )
+  {
+    QgsPointXY center = mSettings.mapToLayerCoordinates( layer, rect.center() );
+    QgsRectangle extentRect = mSettings.mapToLayerCoordinates( layer, extent() );
+    QgsFeatureRequest req = QgsFeatureRequest().setFilterRect( extentRect ).setLimit( 1000 ).setNoAttributes();
```
Comment (borysiasty) [inline src/gui/qgsmapcanvas.cpp]: QgsFeatureRequest::setNoAttributes() ?

Comment (3nids) [reply src/gui/qgsmapcanvas.cpp]: https://github.com/qgis/QGIS/commit/699d8487dc98616bf76dbac409b5a5ba7ab8fcfb

Comment (borysiasty) [reply src/gui/qgsmapcanvas.cpp]: Ah, thanks! :)

File: src/gui/qgsmapcanvas.cpp [hunk 2]
```diff
@@ -990,8 +990,39 @@ void QgsMapCanvas::zoomToSelected( QgsVectorLayer *layer )
   }
 
   rect = mapSettings().layerExtentToOutputExtent( layer, rect );
+
+  // zoom in if point cannot be distinguish from others
+  // also check that rect is empty, as it might not in case of multi points
```
Comment (m-kuhn) [inline src/gui/qgsmapcanvas.cpp]: Thanks for the comment, was just wondering for multipoint layers with singlepoint features.

**Review by m-kuhn — APPROVED** [review]
Looks good to me.

The only question that came to my mind was that there might be different results calculating this in layer coordinates and map coordinates. But this sounds like a lot of effort for a dubious "improvement".

File: src/core/geometry/qgsrectangle.cpp [hunk 3]
```diff
@@ -58,6 +58,13 @@ QgsRectangle QgsRectangle::fromCenterAndSize( QgsPointXY center, double width, d
   return QgsRectangle( xMin, yMin, xMax, yMax );
 }
 
+QgsRectangle QgsRectangle::scaled( double scaleFactor, const QgsPointXY *center ) const
```
Comment (nyalldawson) [inline src/core/geometry/qgsrectangle.cpp]: Unit tests for this please

File: src/gui/qgsmapcanvas.cpp [hunk 4]
```diff
@@ -990,8 +990,40 @@ void QgsMapCanvas::zoomToSelected( QgsVectorLayer *layer )
   }
 
   rect = mapSettings().layerExtentToOutputExtent( layer, rect );
+
+  // zoom in if point cannot be distinguish from others
```
Comment (DelazJ) [inline src/gui/qgsmapcanvas.cpp]: distinguished?

File: src/core/geometry/qgsrectangle.cpp (hunk 3 above)
Comment (3nids) [reply src/core/geometry/qgsrectangle.cpp]: done and for scale too ...

File: src/gui/qgsmapcanvas.cpp (hunk 4 above)
Comment (3nids) [reply src/gui/qgsmapcanvas.cpp]: fixed, thanks!

File: src/core/geometry/qgsrectangle.cpp (hunk 3 above)
Comment (nyalldawson) [reply src/core/geometry/qgsrectangle.cpp]: Thanks!

**Pull request merged**